File: src/questions/parkinglot/java/entity/ParkingFloor.java
package questions.parkinglot.java.entity;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class ParkingFloor {
    private final int floorNumber;
    private final List<ParkingSpot> parkingSpots;

    public ParkingFloor(int floorNumber, Map<VehicleSize, Integer> parkingSpots) {
        this.floorNumber = floorNumber;
        this.parkingSpots = new ArrayList<>();
        initializeSpots(parkingSpots);
        System.out.println("Initialized Floor " + floorNumber + " with spots: " + this.parkingSpots.size());
    }

    private void initializeSpots(Map<VehicleSize, Integer> parkingSpots) {
        for (VehicleSize size : VehicleSize.values()) {
            Integer count = parkingSpots.getOrDefault(size, 0);
            String sizePrefix = getSizePrefix(size);
            for (int i = 1; i <= count; i++) {
                String spotId = String.format("F%d-%s%03d", floorNumber, sizePrefix, i);
                this.parkingSpots.add(new ParkingSpot(spotId, size));
            }
        }
    }

    private String getSizePrefix(VehicleSize size) {
        return switch (size) {
            case SMALL -> "S";
            case MEDIUM -> "M";
            case LARGE -> "L";
            default -> "X";
        };
    }

    public ParkingSpot findAvailableParkingSpot(VehicleSize size) {
        for (ParkingSpot parkingSpot : this.parkingSpots) {
            if (parkingSpot.isSpotAvailable() && parkingSpot.canVehicleFit(size)) {
                return parkingSpot;
            }
        }
        return null;
    }

    public int getAvailableSpotCountBySize(VehicleSize size) {
        int count = 0;
        for (ParkingSpot parkingSpot : this.parkingSpots) {
            if (parkingSpot.isSpotAvailable() && parkingSpot.canVehicleFit(size)) {
                count++;
            }
        }
        return count;
    }

    public int getFloorNumber() {
        return this.floorNumber;
    }

    public List<ParkingSpot> getParkingSpots() {
        return Collections.unmodifiableList(parkingSpots);
    }

    public void displayAvailability() {
        System.out.printf("--- Floor %d Availability ---\n", floorNumber);
        Map<VehicleSize, Long> availableCounts = parkingSpots.stream()
                .filter(ParkingSpot::isSpotAvailable)
                .collect(Collectors.groupingBy(ParkingSpot::getSpotSize, Collectors.counting()));

        for (VehicleSize size : VehicleSize.values()) {
            System.out.printf("  %s spots: %d\n", size, availableCounts.getOrDefault(size, 0L));
        }
    }

}

File: src/questions/parkinglot/py/entity/ParkingFloor.py
from typing import Dict,List
from questions.py.entity.VehicleSize import VehicleSize as VehicleSizeType
from questions.py.entity import VehicleSize
from questions.py.entity.ParkingSpot import ParkingSpot

class ParkingFloor:
    def __init__(self, floor_number: int, spot_counts : Dict[VehicleSizeType,int]):
        self._floor_number = floor_number
        self._spot_counts:List[ParkingSpot] = []
        
    def _initialize_spots(self, spot_counts: Dict[VehicleSize, int]) -> None :
        for vehicle_size in VehicleSize:
            count = spot_counts.get(vehicle_size,0)
            prefix = self._get_size_prefix(vehicle_size)
            for i in range(1,count+1):
                spot_id = f"F{self._floor_number}-{prefix}{i::03d}"
                self._spot_counts.append(ParkingSpot(spot_id,vehicle_size))

    def _get_size_prefix(self, vehicle_size: VehicleSizeType):
        prefix_mapping : Dict[VehicleSize,str] = {
            VehicleSize.SMALL : 'S',
            VehicleSize.MEDIUM : 'M',
            VehicleSize.LARGE : 'L'
        }
        return prefix_mapping.get(vehicle_size,"X")

    def
